File: client/src/data/string-order.js
import casefold from '../../../shared/fixtures/casefold.json' with { type: 'json' };

const folds = casefold.mappings ?? casefold;
const digit = value => value >= '0' && value <= '9';
const failure = message => Object.assign(new Error(message), { code: 'invalid_order', status: 422 });

export function compareCodepoints(left, right) {
  const a = Array.from(left.normalize('NFC')), b = Array.from(right.normalize('NFC'));
  for (let index = 0; index < Math.min(a.length, b.length); index++) {
    const result = a[index].codePointAt(0) - b[index].codePointAt(0);
    if (result) return Math.sign(result);
  }
  return Math.sign(a.length - b.length);
}

export function normalizeStringOrder(input = {}, definitionVersion = 1) {
  if (![1, 2].includes(definitionVersion)) throw failure('Ordering requires definition version 1 or 2');
  if (!input || typeof input !== 'object' || Array.isArray(input) || Object.keys(input).some(key => !['order', 'caseSensitive'].includes(key))) throw failure('Unknown string ordering option');
  if (definitionVersion === 1 && Object.keys(input).length) throw failure('String ordering options require definition version 2');
  if (input.order !== undefined && !['codepoint', 'natural'].includes(input.order)) throw failure('String order must be codepoint or natural');
  if (input.caseSensitive !== undefined && typeof input.caseSensitive !== 'boolean') throw failure('String ordering caseSensitive must be boolean');
  return { order: input.order ?? 'codepoint', caseSensitive: input.caseSensitive ?? true };
}

function comparable(value, caseSensitive) {
  const normalized = value.normalize('NFC');
  return caseSensitive ? normalized : Array.from(normalized, char => folds[String(char.codePointAt(0))] ?? char).join('').normalize('NFC');
}

function naturalCompare(left, right) {
  const a = Array.from(left), b = Array.from(right);
  let i = 0, j = 0;
  while (i < a.length && j < b.length) {
    if (digit(a[i]) && digit(b[j])) {
      const startA = i, startB = j;
      while (i < a.length && digit(a[i])) i++;
      while (j < b.length && digit(b[j])) j++;
      let significantA = startA, significantB = startB;
      while (significantA < i - 1 && a[significantA] === '0') significantA++;
      while (significantB < j - 1 && b[significantB] === '0') significantB++;
      const magnitude = (i - significantA) - (j - significantB);
      if (magnitude) return Math.sign(magnitude);
      for (let offset = 0; offset < i - significantA; offset++) {
        const difference = a[significantA + offset].codePointAt(0) - b[significantB + offset].codePointAt(0);
        if (difference) return Math.sign(difference);
      }
      const length = (i - startA) - (j - startB);
      if (length) return Math.sign(length);
    } else {
      const difference = a[i].codePointAt(0) - b[j].codePointAt(0);
      if (difference) return Math.sign(difference);
      i++; j++;
    }
  }
  return Math.sign((a.length - i) - (b.length - j));
}

export function compareOrderedText(left, right, { order = 'codepoint', caseSensitive = true } = {}) {
  const a = comparable(left, caseSensitive), b = comparable(right, caseSensitive);
  const result = order === 'natural' ? naturalCompare(a, b) : compareCodepoints(a, b);
  return result || compareCodepoints(left, right);
}

File: server/app/services/string_order.py
"""Locale-independent, versioned string ordering using the shared casefold table."""
from __future__ import annotations

import unicodedata

from ..models.domain import DomainError, folded


def normalize_string_order(value, definition_version=1):
    if type(definition_version) is not int or definition_version not in (1, 2):
        raise DomainError('invalid_order', 'Ordering requires definition version 1 or 2.')
    if not isinstance(value, dict) or set(value) - {'order', 'caseSensitive'}:
        raise DomainError('invalid_order', 'Unknown string ordering option.')
    if definition_version == 1 and value:
        raise DomainError('invalid_order', 'String ordering options require definition version 2.')
    if value.get('order', 'codepoint') not in ('codepoint', 'natural'):
        raise DomainError('invalid_order', 'String order must be codepoint or natural.')
    if type(value.get('caseSensitive', True)) is not bool:
        raise DomainError('invalid_order', 'String ordering caseSensitive must be boolean.')
    return {'order': value.get('order', 'codepoint'), 'caseSensitive': value.get('caseSensitive', True)}


def _compare(left, right):
    return (left > right) - (left < right)


def _natural_compare(left, right):
    i = j = 0
    while i < len(left) and j < len(right):
        if '0' <= left[i] <= '9' and '0' <= right[j] <= '9':
            start_left, start_right = i, j
            while i < len(left) and '0' <= left[i] <= '9':
                i += 1
            while j < len(right) and '0' <= right[j] <= '9':
                j += 1
            # Compare magnitude by significant length and digits, never float/int conversion.
            a = left[start_left:i].lstrip('0') or '0'
            b = right[start_right:j].lstrip('0') or '0'
            result = _compare(len(a), len(b)) or _compare(a, b) or _compare(i - start_left, j - start_right)
            if result:
                return result
        else:
            result = _compare(left[i], right[j])
            if result:
                return result
            i += 1
            j += 1
    return _compare(len(left) - i, len(right) - j)


def compare_ordered_text(left, right, options=None):
    options = {} if options is None else options
    original_left, original_right = unicodedata.normalize('NFC', left), unicodedata.normalize('NFC', right)
    a, b = (original_left, original_right) if options.get('caseSensitive', True) else (folded(left), folded(right))
    result = _natural_compare(a, b) if options.get('order', 'codepoint') == 'natural' else _compare(a, b)
    return result or _compare(original_left, original_right)
